### PR TITLE
Use inline type hints in `handlers/` and `rest/`

### PR DESCRIPTION
**Pull request by ShadowJonathan**
Part of #8351

This PR is tantamount to running

```
python3.8 -m com2ann -v 6 handlers/
python3.8 -m com2ann -v 6 rest/
```

(com2ann requires python 3.8 to run)

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

---

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: synapse/handlers/_base.py [hunk 1]
```diff
@@ -38,10 +38,10 @@ class BaseHandler:
     """
 
     def __init__(self, hs: "HomeServer"):
-        self.store = hs.get_datastore()  # type: synapse.storage.DataStore
+        self.store: synapse.storage.DataStore = hs.get_datastore()
```
Comment (clokep) [inline synapse/handlers/_base.py:41]: Are...these even necessary? Now that `hs` has a type hint I think we should get these automatically? (Same for `state_handler` below).

File: synapse/handlers/oidc.py [hunk 2]
```diff
@@ -301,11 +301,11 @@ def __init__(
                 hs.get_clock(),
             )
 
-        self._client_auth = ClientAuth(
+        self._client_auth: ClientAuth = ClientAuth(
```
Comment (clokep) [inline synapse/handlers/oidc.py:304]: I think we can just remove this type hint?

File: synapse/rest/media/v1/media_repository.py [hunk 3]
```diff
@@ -78,16 +78,16 @@ def __init__(self, hs: "HomeServer"):
 
         Thumbnailer.set_limits(self.max_image_pixels)
 
-        self.primary_base_path = hs.config.media_store_path  # type: str
-        self.filepaths = MediaFilePaths(self.primary_base_path)  # type: MediaFilePaths
+        self.primary_base_path: str = hs.config.media_store_path
+        self.filepaths: MediaFilePaths = MediaFilePaths(self.primary_base_path)
```
Comment (clokep) [inline synapse/rest/media/v1/media_repository.py:82]: Another spot I'm surprised we need a type hint?

**Review by clokep — APPROVED** [review]
Looks good overall! There's a few spots that I marked which I'm surprised we even need type hints. It's a bit of scope creep though so we can ignore it for now if you'd like. (Maybe there's a mypy setting that lets you find such things?)

File: synapse/handlers/_base.py (hunk 1 above)
Comment (ShadowJonathan) [reply synapse/handlers/_base.py:41]: Just looked at HomeServer, yeah, this is probably not necessary as mypy can infer this directly from the `get_datastore` call signature, do you want me to remove it?

File: synapse/handlers/oidc.py (hunk 2 above)
Comment (ShadowJonathan) [reply synapse/handlers/oidc.py:304]: Oh wow, yeah, i'll remove this. :+1:

File: synapse/rest/media/v1/media_repository.py (hunk 3 above)
Comment (ShadowJonathan) [reply synapse/rest/media/v1/media_repository.py:82]: Yeah they're automated changes, i'm more surprised that they're assigned there in the first place though. I'll remove :+1:

File: synapse/handlers/_base.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/_base.py:41]: If it isn't necessary than yep! It is probably from before we had a `hs: "HomeServer"` type hint or something?

Comment (ShadowJonathan) [reply synapse/handlers/_base.py:41]: Probably, i'll have to update the "how to update instance variable type annotations" guidelines on #8351 somewhat to reflect this, but i'll remove this :+1:

File: synapse/rest/media/v1/media_repository.py [hunk 4]
```diff
@@ -79,7 +79,7 @@ def __init__(self, hs: "HomeServer"):
         Thumbnailer.set_limits(self.max_image_pixels)
 
         self.primary_base_path: str = hs.config.media_store_path
-        self.filepaths = MediaFilePaths(self.primary_base_path)
+        self.filepaths: MediaFilePaths = MediaFilePaths(self.primary_base_path)
```
Comment (ShadowJonathan) [inline synapse/rest/media/v1/media_repository.py:82]: For some reason mypy wasn't able to infer this after i removed it :/

Comment (clokep) [reply synapse/rest/media/v1/media_repository.py:82]: That's odd, but let's just leave it for now.

**Review by clokep — APPROVED** [review]

**Pull request merged**